unable to spoon RxJava

```
Caused by: java.lang.NullPointerException
	at spoon.support.compiler.jdt.JDTTreeBuilder.createType(JDTTreeBuilder.java:946)
	at spoon.support.compiler.jdt.JDTTreeBuilder.visit(JDTTreeBuilder.java:2972)
	at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.traverse(TypeDeclaration.java:1374)
	at spoon.support.compiler.jdt.JDTTreeBuilder.visit(JDTTreeBuilder.java:2511)
	at org.eclipse.jdt.internal.compiler.ast.QualifiedAllocationExpression.traverse(QualifiedAllocationExpression.java:604)
	at spoon.support.compiler.jdt.JDTTreeBuilder.visit(JDTTreeBuilder.java:2315)
	at org.eclipse.jdt.internal.compiler.ast.MessageSend.traverse(MessageSend.java:1031)
	at org.eclipse.jdt.internal.compiler.ast.Block.traverse(Block.java:139)
	at spoon.support.compiler.jdt.JDTTreeBuilder.visit(JDTTreeBuilder.java:1602)
	at org.eclipse.jdt.internal.compiler.ast.LambdaExpression.traverse(LambdaExpression.java:678)
	at spoon.support.compiler.jdt.JDTTreeBuilder.visit(JDTTreeBuilder.java:2315)
	at org.eclipse.jdt.internal.compiler.ast.MessageSend.traverse(MessageSend.java:1031)
	at spoon.support.compiler.jdt.JDTTreeBuilder.visit(JDTTreeBuilder.java:2219)
	at org.eclipse.jdt.internal.compiler.ast.LocalDeclaration.traverse(LocalDeclaration.java:301)
	at spoon.support.compiler.jdt.JDTTreeBuilder.visit(JDTTreeBuilder.java:2396)
	at org.eclipse.jdt.internal.compiler.ast.MethodDeclaration.traverse(MethodDeclaration.java:327)
	at spoon.support.compiler.jdt.JDTTreeBuilder.visit(JDTTreeBuilder.java:3053)
	at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.traverse(TypeDeclaration.java:1319)
	at org.eclipse.jdt.internal.compiler.ast.CompilationUnitDeclaration.traverse(CompilationUnitDeclaration.java:748)
	at org.eclipse.jdt.internal.compiler.ast.CompilationUnitDeclaration.traverse(CompilationUnitDeclaration.java:709)
	at spoon.support.compiler.jdt.JDTBasedSpoonCompiler.buildSources(JDTBasedSpoonCompiler.java:227)
	at spoon.support.compiler.jdt.JDTBasedSpoonCompiler.build(JDTBasedSpoonCompiler.java:480)
	at spoon.Launcher.buildModel(Launcher.java:664)
	at spoon.Launcher.run(Launcher.java:616)
	at spoon.SpoonAPI$run$1.call(Unknown Source)
```
As far as I can tell the logging is not granular enough to narrow down the scope of the problem. I'm trying to analyze the [RxJava 2.x sources][1] (note the `2.x` branch). Unfortunately there's quite a lot of classes and they are interdependent so I've been unable to find the cause.

I was forced onto 4.4.0-SNAPSHOT as when attempting this with 4.3.0 another error occurred:
```
Caused by: java.lang.RuntimeException: Unknown TypeBinding: class org.eclipse.jdt.internal.compiler.lookup.IntersectionCastTypeBinding capture#776-of ? super T & capture#778-of ? super T
	at spoon.support.compiler.jdt.JDTTreeBuilder$ReferenceBuilder.getTypeReference(JDTTreeBuilder.java:617)
	at spoon.support.compiler.jdt.JDTTreeBuilder$ReferenceBuilder.getTypeReference(JDTTreeBuilder.java:461)
	at spoon.support.compiler.jdt.JDTTreeBuilder.buildCommonPartForCtNewClassAndCtConstructorCall(JDTTreeBuilder.java:1611)
	at spoon.support.compiler.jdt.JDTTreeBuilder.visit(JDTTreeBuilder.java:1592)
	at org.eclipse.jdt.internal.compiler.ast.AllocationExpression.traverse(AllocationExpression.java:657)
	at spoon.support.compiler.jdt.JDTTreeBuilder.visit(JDTTreeBuilder.java:2279)
	at org.eclipse.jdt.internal.compiler.ast.MessageSend.traverse(MessageSend.java:1008)
	at org.eclipse.jdt.internal.compiler.ast.ReturnStatement.traverse(ReturnStatement.java:373)
	at spoon.support.compiler.jdt.JDTTreeBuilder.visit(JDTTreeBuilder.java:2360)
	at org.eclipse.jdt.internal.compiler.ast.MethodDeclaration.traverse(MethodDeclaration.java:317)
	at spoon.support.compiler.jdt.JDTTreeBuilder.visit(JDTTreeBuilder.java:3024)
	at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.traverse(TypeDeclaration.java:1338)
	at org.eclipse.jdt.internal.compiler.ast.CompilationUnitDeclaration.traverse(CompilationUnitDeclaration.java:748)
	at org.eclipse.jdt.internal.compiler.ast.CompilationUnitDeclaration.traverse(CompilationUnitDeclaration.java:709)
	at spoon.support.compiler.jdt.JDTBasedSpoonCompiler.buildSources(JDTBasedSpoonCompiler.java:247)
	at spoon.support.compiler.jdt.JDTBasedSpoonCompiler.build(JDTBasedSpoonCompiler.java:500)
	at spoon.Launcher.buildModel(Launcher.java:748)
	at spoon.Launcher.run(Launcher.java:700)
	at spoon.SpoonAPI$run$1.call(Unknown Source)
```
Perhaps that provides some insight into what the cause might be.

 [1]: https://github.com/ReactiveX/RxJava/tree/2.x

--- a/src/main/java/spoon/reflect/factory/Factory.java
+++ b/src/main/java/spoon/reflect/factory/Factory.java
@@ -40,4 +40,6 @@ public interface Factory {
 	EvalFactory Eval(); // used 4 times
 
 	ConstructorFactory Constructor(); // used 3 times
+
+	InternalFactory Internal();
 }

--- a/src/main/java/spoon/reflect/factory/FactoryImpl.java
+++ b/src/main/java/spoon/reflect/factory/FactoryImpl.java
@@ -17,11 +17,6 @@
 
 package spoon.reflect.factory;
 
-import java.io.Serializable;
-import java.util.HashMap;
-import java.util.Map;
-import java.util.Random;
-
 import spoon.compiler.Environment;
 import spoon.reflect.cu.CompilationUnit;
 import spoon.reflect.declaration.CtAnnotationType;
@@ -35,7 +30,13 @@ import spoon.reflect.declaration.CtMethod;
 import spoon.reflect.declaration.CtPackage;
 import spoon.reflect.declaration.CtType;
 import spoon.support.DefaultCoreFactory;
+import spoon.support.DefaultInternalFactory;
 import spoon.support.StandardEnvironment;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
 
 /**
  * Implements {@link Factory}
@@ -251,6 +252,16 @@ public class FactoryImpl implements Factory, Serializable {
 			type = new TypeFactory(this);
 		}
 		return type;
+	}
+
+	private transient InternalFactory internal;
+
+	@Override
+	public InternalFactory Internal() {
+		if (internal == null) {
+			internal = new DefaultInternalFactory(this);
+		}
+		return internal;
 	}
 
 	/**

--- a/src/main/java/spoon/reflect/factory/InternalFactory.java
+++ b/src/main/java/spoon/reflect/factory/InternalFactory.java
@@ -1,0 +1,15 @@
+package spoon.reflect.factory;
+
+import spoon.reflect.internal.CtCircularTypeReference;
+
+/**
+ * This interface defines the creation methods for internal nodes of the
+ * meta-model. These nodes are available in the AST provided by Spoon
+ * but their creation should be used only in internal of Spoon.
+ */
+public interface InternalFactory {
+	/**
+	 * Creates a circular type reference.
+	 */
+	<T> CtCircularTypeReference createCircularTypeReference();
+}

--- a/src/main/java/spoon/reflect/internal/CtCircularTypeReference.java
+++ b/src/main/java/spoon/reflect/internal/CtCircularTypeReference.java
@@ -1,0 +1,16 @@
+package spoon.reflect.internal;
+
+import spoon.reflect.reference.CtTypeParameterReference;
+import spoon.reflect.reference.CtTypeReference;
+
+/**
+ * When we build a {@link CtTypeReference}, we can have a circular when
+ * we got this kind of generic type: {@code <T extends Comparable<? super T>>}.
+ * In this case, where we are at the last T, we come back at the first
+ * one and we are in a circular.
+ *
+ * Now, the last T is a CtCircularTypeReference and we stop the circular
+ * when we build the generic or when we scan an AST given.
+ */
+public interface CtCircularTypeReference extends CtTypeParameterReference {
+}

--- a/src/main/java/spoon/reflect/visitor/CtAbstractVisitor.java
+++ b/src/main/java/spoon/reflect/visitor/CtAbstractVisitor.java
@@ -63,6 +63,7 @@ import spoon.reflect.declaration.CtParameter;
 import spoon.reflect.declaration.CtTypeParameter;
 import spoon.reflect.reference.CtArrayTypeReference;
 import spoon.reflect.reference.CtCatchVariableReference;
+import spoon.reflect.internal.CtCircularTypeReference;
 import spoon.reflect.reference.CtExecutableReference;
 import spoon.reflect.reference.CtFieldReference;
 import spoon.reflect.reference.CtLocalVariableReference;
@@ -385,6 +386,11 @@ public abstract class CtAbstractVisitor implements CtVisitor {
 
 	@Override
 	public <T> void visitCtTypeReference(CtTypeReference<T> reference) {
+
+	}
+
+	@Override
+	public <T> void visitCtCircularTypeReference(CtCircularTypeReference reference) {
 
 	}
 

--- a/src/main/java/spoon/reflect/visitor/CtInheritanceScanner.java
+++ b/src/main/java/spoon/reflect/visitor/CtInheritanceScanner.java
@@ -97,6 +97,7 @@ import spoon.reflect.declaration.CtTypedElement;
 import spoon.reflect.declaration.CtVariable;
 import spoon.reflect.reference.CtArrayTypeReference;
 import spoon.reflect.reference.CtCatchVariableReference;
+import spoon.reflect.internal.CtCircularTypeReference;
 import spoon.reflect.reference.CtExecutableReference;
 import spoon.reflect.reference.CtFieldReference;
 import spoon.reflect.reference.CtGenericElementReference;
@@ -761,6 +762,11 @@ public abstract class CtInheritanceScanner implements CtVisitor {
 		scanCtGenericElementReference(e);
 		scanCtTypeAnnotableReference(e);
 		scanCtVisitable(e);
+	}
+
+	@Override
+	public <T> void visitCtCircularTypeReference(CtCircularTypeReference e) {
+		visitCtTypeParameterReference(e);
 	}
 
 	@Override

--- a/src/main/java/spoon/reflect/visitor/CtScanner.java
+++ b/src/main/java/spoon/reflect/visitor/CtScanner.java
@@ -17,9 +17,6 @@
 
 package spoon.reflect.visitor;
 
-import java.lang.annotation.Annotation;
-import java.util.Collection;
-
 import spoon.reflect.code.CtAnnotationFieldAccess;
 import spoon.reflect.code.CtArrayAccess;
 import spoon.reflect.code.CtArrayRead;
@@ -65,8 +62,8 @@ import spoon.reflect.code.CtTry;
 import spoon.reflect.code.CtTryWithResource;
 import spoon.reflect.code.CtTypeAccess;
 import spoon.reflect.code.CtUnaryOperator;
-import spoon.reflect.code.CtVariableRead;
 import spoon.reflect.code.CtVariableAccess;
+import spoon.reflect.code.CtVariableRead;
 import spoon.reflect.code.CtVariableWrite;
 import spoon.reflect.code.CtWhile;
 import spoon.reflect.declaration.CtAnnotation;
@@ -84,6 +81,7 @@ import spoon.reflect.declaration.CtParameter;
 import spoon.reflect.declaration.CtTypeParameter;
 import spoon.reflect.reference.CtArrayTypeReference;
 import spoon.reflect.reference.CtCatchVariableReference;
+import spoon.reflect.internal.CtCircularTypeReference;
 import spoon.reflect.reference.CtExecutableReference;
 import spoon.reflect.reference.CtFieldReference;
 import spoon.reflect.reference.CtLocalVariableReference;
@@ -93,6 +91,9 @@ import spoon.reflect.reference.CtReference;
 import spoon.reflect.reference.CtTypeParameterReference;
 import spoon.reflect.reference.CtTypeReference;
 import spoon.reflect.reference.CtUnboundVariableReference;
+
+import java.lang.annotation.Annotation;
+import java.util.Collection;
 
 /**
  * This visitor implements a deep-search scan on the metamodel.
@@ -681,6 +682,10 @@ public abstract class CtScanner implements CtVisitor {
 		scan(reference.getDeclaringType());
 		scanReferences(reference.getActualTypeArguments());
 		exitReference(reference);
+	}
+
+	@Override
+	public <T> void visitCtCircularTypeReference(CtCircularTypeReference reference) {
 	}
 
 	@Override

--- a/src/main/java/spoon/reflect/visitor/CtVisitor.java
+++ b/src/main/java/spoon/reflect/visitor/CtVisitor.java
@@ -79,6 +79,7 @@ import spoon.reflect.declaration.CtParameter;
 import spoon.reflect.declaration.CtTypeParameter;
 import spoon.reflect.reference.CtArrayTypeReference;
 import spoon.reflect.reference.CtCatchVariableReference;
+import spoon.reflect.internal.CtCircularTypeReference;
 import spoon.reflect.reference.CtExecutableReference;
 import spoon.reflect.reference.CtFieldReference;
 import spoon.reflect.reference.CtLocalVariableReference;
@@ -387,6 +388,11 @@ public interface CtVisitor {
 	 * Visits a reference to a type.
 	 */
 	<T> void visitCtTypeReference(CtTypeReference<T> reference);
+
+	/**
+	 * Visits a circular reference.
+	 */
+	<T> void visitCtCircularTypeReference(CtCircularTypeReference reference);
 
 	/**
 	 * Visits a type access.

--- a/src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java
+++ b/src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java
@@ -17,16 +17,6 @@
 
 package spoon.reflect.visitor;
 
-import java.lang.annotation.Annotation;
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.Map.Entry;
-import java.util.Stack;
-
 import org.apache.log4j.Level;
 import spoon.compiler.Environment;
 import spoon.reflect.code.BinaryOperatorKind;
@@ -77,8 +67,8 @@ import spoon.reflect.code.CtTry;
 import spoon.reflect.code.CtTryWithResource;
 import spoon.reflect.code.CtTypeAccess;
 import spoon.reflect.code.CtUnaryOperator;
-import spoon.reflect.code.CtVariableRead;
 import spoon.reflect.code.CtVariableAccess;
+import spoon.reflect.code.CtVariableRead;
 import spoon.reflect.code.CtVariableWrite;
 import spoon.reflect.code.CtWhile;
 import spoon.reflect.code.UnaryOperatorKind;
@@ -105,6 +95,7 @@ import spoon.reflect.declaration.ParentNotInitializedException;
 import spoon.reflect.factory.Factory;
 import spoon.reflect.reference.CtArrayTypeReference;
 import spoon.reflect.reference.CtCatchVariableReference;
+import spoon.reflect.internal.CtCircularTypeReference;
 import spoon.reflect.reference.CtExecutableReference;
 import spoon.reflect.reference.CtFieldReference;
 import spoon.reflect.reference.CtGenericElementReference;
@@ -119,6 +110,16 @@ import spoon.reflect.reference.CtUnboundVariableReference;
 import spoon.support.reflect.cu.CtLineElementComparator;
 import spoon.support.util.SortedList;
 import spoon.support.visitor.SignaturePrinter;
+
+import java.lang.annotation.Annotation;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Stack;
 
 /**
  * A visitor for generating Java code from the program compile-time model.
@@ -1826,6 +1827,11 @@ public class DefaultJavaPrettyPrinter implements CtVisitor, PrettyPrinter {
 		if (!context.ignoreGenerics) {
 			writeActualTypeArguments(ref);
 		}
+	}
+
+	@Override
+	public <T> void visitCtCircularTypeReference(CtCircularTypeReference reference) {
+		visitCtTypeReference(reference);
 	}
 
 	private <T> boolean hasDeclaringTypeWithGenerics(CtTypeReference<T> reference) {

--- a/src/main/java/spoon/support/DefaultInternalFactory.java
+++ b/src/main/java/spoon/support/DefaultInternalFactory.java
@@ -1,0 +1,21 @@
+package spoon.support;
+
+import spoon.reflect.factory.Factory;
+import spoon.reflect.factory.InternalFactory;
+import spoon.reflect.internal.CtCircularTypeReference;
+import spoon.support.reflect.internal.CtCircularTypeReferenceImpl;
+
+public class DefaultInternalFactory implements InternalFactory {
+	Factory mainFactory;
+
+	public DefaultInternalFactory(Factory factory) {
+		mainFactory = factory;
+	}
+
+	@Override
+	public <T> CtCircularTypeReference createCircularTypeReference() {
+		CtCircularTypeReference e = new CtCircularTypeReferenceImpl();
+		e.setFactory(mainFactory);
+		return e;
+	}
+}

--- a/src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java
+++ b/src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java
@@ -17,19 +17,6 @@
 
 package spoon.support.compiler.jdt;
 
-import static spoon.reflect.ModelElementContainerDefaultCapacities.CASTS_CONTAINER_DEFAULT_CAPACITY;
-
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.EnumSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
-import java.util.Stack;
-import java.util.TreeMap;
-import java.util.TreeSet;
-
 import org.apache.log4j.Logger;
 import org.eclipse.jdt.core.compiler.CharOperation;
 import org.eclipse.jdt.internal.compiler.ASTVisitor;
@@ -151,7 +138,6 @@ import org.eclipse.jdt.internal.compiler.lookup.TypeBinding;
 import org.eclipse.jdt.internal.compiler.lookup.TypeVariableBinding;
 import org.eclipse.jdt.internal.compiler.lookup.VariableBinding;
 import org.eclipse.jdt.internal.compiler.lookup.WildcardBinding;
-
 import spoon.reflect.code.BinaryOperatorKind;
 import spoon.reflect.code.CtAnnotationFieldAccess;
 import spoon.reflect.code.CtArrayAccess;
@@ -213,6 +199,7 @@ import spoon.reflect.factory.CoreFactory;
 import spoon.reflect.factory.Factory;
 import spoon.reflect.reference.CtArrayTypeReference;
 import spoon.reflect.reference.CtCatchVariableReference;
+import spoon.reflect.internal.CtCircularTypeReference;
 import spoon.reflect.reference.CtExecutableReference;
 import spoon.reflect.reference.CtFieldReference;
 import spoon.reflect.reference.CtLocalVariableReference;
@@ -224,6 +211,20 @@ import spoon.reflect.reference.CtVariableReference;
 import spoon.reflect.visitor.Query;
 import spoon.reflect.visitor.filter.TypeFilter;
 import spoon.support.reflect.reference.CtUnboundVariableReferenceImpl;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+import static spoon.reflect.ModelElementContainerDefaultCapacities.CASTS_CONTAINER_DEFAULT_CAPACITY;
 
 /**
  * A visitor for iterating through the parse tree.
@@ -439,8 +440,7 @@ public class JDTTreeBuilder extends ASTVisitor {
 			return ref;
 		}
 
-		// Map<TypeBinding, CtTypeReference<?>> bindingCache = new
-		// HashMap<TypeBinding, CtTypeReference<?>>();
+		final Map<TypeBinding, CtTypeReference> bindingCache = new HashMap<TypeBinding, CtTypeReference>();
 
 		public <T> CtTypeReference<T> getTypeReference(TypeBinding binding, TypeReference ref) {
 			CtTypeReference<T> ctRef = getTypeReference(binding);
@@ -544,7 +544,7 @@ public class JDTTreeBuilder extends ASTVisitor {
 				}
 				if (bounds && b.superInterfaces != null && b.superInterfaces != Binding.NO_SUPERINTERFACES) {
 					bounds = false;
-					// bindingCache.put(binding, ref);
+					bindingCache.put(binding, ref);
 					for (int i = 0, length = b.superInterfaces.length; i < length; i++) {
 						TypeBinding tb = b.superInterfaces[i];
 						((CtTypeParameterReference) ref).addBound(getTypeReference(tb));
@@ -574,7 +574,18 @@ public class JDTTreeBuilder extends ASTVisitor {
 				}
 
 				if (((WildcardBinding) binding).bound != null && ref instanceof CtTypeParameterReference) {
-					((CtTypeParameterReference) ref).addBound(getTypeReference(((WildcardBinding) binding).bound));
+					if (bindingCache.containsKey(((WildcardBinding) binding).bound)) {
+						final CtCircularTypeReference circularRef = factory.Internal().createCircularTypeReference();
+						final CtTypeReference originalRef = bindingCache.get(((WildcardBinding) binding).bound);
+						circularRef.setPackage(originalRef.getPackage());
+						circularRef.setSimpleName(originalRef.getSimpleName());
+						circularRef.setDeclaringType(originalRef.getDeclaringType());
+						circularRef.setActualTypeArguments(originalRef.getActualTypeArguments());
+						circularRef.setTypeAnnotations(originalRef.getTypeAnnotations());
+						((CtTypeParameterReference) ref).addBound(circularRef);
+					} else {
+						((CtTypeParameterReference) ref).addBound(getTypeReference(((WildcardBinding) binding).bound));
+					}
 				}
 			} else if (binding instanceof LocalTypeBinding) {
 				if (!JDTTreeBuilder.this.context.isGenericTypeExplicit) {
@@ -639,7 +650,7 @@ public class JDTTreeBuilder extends ASTVisitor {
 			} else {
 				throw new RuntimeException("Unknown TypeBinding: " + binding.getClass() + " " + binding);
 			}
-			// bindingCache.remove(binding);
+			bindingCache.remove(binding);
 			addTypeAnnotationFromBindingToReference(binding, ref);
 			return (CtTypeReference<T>) ref;
 		}

--- a/src/main/java/spoon/support/reflect/eval/VisitorPartialEvaluator.java
+++ b/src/main/java/spoon/support/reflect/eval/VisitorPartialEvaluator.java
@@ -90,6 +90,7 @@ import spoon.reflect.declaration.ModifierKind;
 import spoon.reflect.eval.PartialEvaluator;
 import spoon.reflect.reference.CtArrayTypeReference;
 import spoon.reflect.reference.CtCatchVariableReference;
+import spoon.reflect.internal.CtCircularTypeReference;
 import spoon.reflect.reference.CtExecutableReference;
 import spoon.reflect.reference.CtFieldReference;
 import spoon.reflect.reference.CtGenericElementReference;
@@ -717,6 +718,11 @@ public class VisitorPartialEvaluator implements CtVisitor, PartialEvaluator {
 	}
 
 	public <T> void visitCtTypeReference(CtTypeReference<T> reference) {
+		throw new RuntimeException("Unknow Element");
+	}
+
+	@Override
+	public <T> void visitCtCircularTypeReference(CtCircularTypeReference reference) {
 		throw new RuntimeException("Unknow Element");
 	}
 

--- a/src/main/java/spoon/support/reflect/internal/CtCircularTypeReferenceImpl.java
+++ b/src/main/java/spoon/support/reflect/internal/CtCircularTypeReferenceImpl.java
@@ -1,0 +1,12 @@
+package spoon.support.reflect.internal;
+
+import spoon.reflect.internal.CtCircularTypeReference;
+import spoon.reflect.visitor.CtVisitor;
+import spoon.support.reflect.reference.CtTypeParameterReferenceImpl;
+
+public class CtCircularTypeReferenceImpl extends CtTypeParameterReferenceImpl implements CtCircularTypeReference {
+	@Override
+	public void accept(CtVisitor visitor) {
+		visitor.visitCtCircularTypeReference(this);
+	}
+}

--- a/src/main/java/spoon/support/visitor/SignaturePrinter.java
+++ b/src/main/java/spoon/support/visitor/SignaturePrinter.java
@@ -82,6 +82,7 @@ import spoon.reflect.declaration.CtParameter;
 import spoon.reflect.declaration.CtTypeParameter;
 import spoon.reflect.reference.CtArrayTypeReference;
 import spoon.reflect.reference.CtCatchVariableReference;
+import spoon.reflect.internal.CtCircularTypeReference;
 import spoon.reflect.reference.CtExecutableReference;
 import spoon.reflect.reference.CtFieldReference;
 import spoon.reflect.reference.CtLocalVariableReference;
@@ -619,6 +620,11 @@ public class SignaturePrinter implements CtVisitor {
 
 	public <T> void visitCtTypeReference(CtTypeReference<T> reference) {
 		write(reference.getQualifiedName());
+	}
+
+	@Override
+	public <T> void visitCtCircularTypeReference(CtCircularTypeReference reference) {
+		visitCtTypeReference(reference);
 	}
 
 	@Override

--- a/src/test/java/spoon/test/reference/TypeReferenceTest.java
+++ b/src/test/java/spoon/test/reference/TypeReferenceTest.java
@@ -5,15 +5,19 @@ import spoon.Launcher;
 import spoon.compiler.SpoonCompiler;
 import spoon.compiler.SpoonResource;
 import spoon.compiler.SpoonResourceHelper;
+import spoon.reflect.code.CtInvocation;
 import spoon.reflect.declaration.CtField;
 import spoon.reflect.declaration.CtInterface;
 import spoon.reflect.declaration.CtType;
 import spoon.reflect.factory.Factory;
+import spoon.reflect.internal.CtCircularTypeReference;
 import spoon.reflect.reference.CtExecutableReference;
 import spoon.reflect.reference.CtFieldReference;
+import spoon.reflect.reference.CtTypeParameterReference;
 import spoon.reflect.reference.CtTypeReference;
 import spoon.reflect.visitor.Query;
 import spoon.reflect.visitor.filter.ReferenceTypeFilter;
+import spoon.reflect.visitor.filter.TypeFilter;
 
 import java.util.Collection;
 import java.util.List;
@@ -21,6 +25,8 @@ import java.util.List;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
 
 /**
  * @author Lionel Seinturier
@@ -178,6 +184,41 @@ public class TypeReferenceTest {
 
 		assertNotEquals(firstRef.toString(), secondRef.toString());
 		assertNotEquals(firstRef, secondRef);
+	}
+
+	@Test
+	public void testRecursiveTypeReference() throws Exception {
+		final Launcher launcher = new Launcher();
+		launcher.addInputResource("./src/test/java/spoon/test/reference/testclasses/Tacos.java");
+		launcher.setSourceOutputDirectory("./target/spoon-test");
+		launcher.run();
+
+		final CtInvocation<?> inv = Query.getElements(launcher.getFactory(), new TypeFilter<CtInvocation<?>>(CtInvocation.class) {
+			@Override
+			public boolean matches(CtInvocation<?> element) {
+				return !element.getExecutable().isConstructor() && super.matches(element);
+			}
+		}).get(0);
+
+		assertNotNull(inv.getExecutable());
+		final CtTypeReference<?> returnType = inv.getExecutable().getType();
+		assertNotNull(returnType);
+		assertEquals(1, returnType.getActualTypeArguments().size());
+
+		final CtTypeParameterReference genericType = (CtTypeParameterReference) returnType.getActualTypeArguments().get(0);
+		assertNotNull(genericType);
+		assertEquals(1, genericType.getBounds().size());
+
+		final CtTypeReference<?> extendsGeneric = genericType.getBounds().get(0);
+		assertNotNull(extendsGeneric);
+		assertEquals(1, extendsGeneric.getActualTypeArguments().size());
+
+		final CtTypeParameterReference genericExtends = (CtTypeParameterReference) extendsGeneric.getActualTypeArguments().get(0);
+		assertNotNull(genericExtends);
+		assertEquals(1, genericExtends.getBounds().size());
+
+		final CtTypeReference<?> circularRef = genericExtends.getBounds().get(0);
+		assertTrue(circularRef instanceof CtCircularTypeReference);
 	}
 
 	class A {

--- a/src/test/java/spoon/test/reference/testclasses/Tacos.java
+++ b/src/test/java/spoon/test/reference/testclasses/Tacos.java
@@ -1,0 +1,10 @@
+package spoon.test.reference.testclasses;
+
+import java.util.Comparator;
+
+public class Tacos {
+	@SuppressWarnings({ "unchecked", "rawtypes"})
+	public final void toSortedList() {
+		Comparator.naturalOrder();
+	}
+}

